internals/secrethub/run.go

```diff
@@ -422,6 +422,48 @@ func ReadEnvFile(filepath string, varReader tpl.VariableReader, parser tpl.Parse
 	}, nil
 }
 
+// OsEnv is an environment with secrets configured with the
+// secrethub:// syntax in the os environment variables.
+type OsEnv struct {
+	envVars map[string]string
+}
+
+// NewOsEnvSource returns an environment with secrets configured in the
+// os environment with the secrethub:// syntax.
+func NewOsEnvSource(osEnv map[string]string) *OsEnv {
+	envVars := make(map[string]string)
+	for key, value := range osEnv {
+		if strings.HasPrefix(value, "secrethub://") {
+			envVars[key] = strings.TrimPrefix(value, "secrethub://")
+		}
+	}
+	return &OsEnv{
+		envVars: envVars,
+	}
+}
+
+// Env returns a map of key value pairs with the secrets configured with the
+// secrethub:// syntax.
+func (env *OsEnv) Env(secrets map[string]string, _ tpl.SecretReader) (map[string]string, error) {
+	envVarsWithSecrets := make(map[string]string)
+	for key, path := range env.envVars {
+		envVarsWithSecrets[key] = secrets[path]
```

Comment:
Should I use the `SecretReader` or the secrets from the map?

Reply:
Right now, with the `--ignore-missing-secrets` flag the environment variables with missing secrets in them are cleared. Is this the expected behavior?

Reply:
> Should I use the SecretReader or the secrets from the map?

I believe you can currently choose; All secret paths returned with `Secrets` are present in the `secrets` map. So you can either return the secret paths in `Secrets` and use the `secrets` map, or return an empty list with `Secrets` and use the secret reader.

I think this was introduced together with template variables. For template variables we can't determine in advance what the secret paths will be (as they dependent on evaluating the template variables). That's when the secret reader was introduced so that the secrets can be read when evaluating the template.

It would be nice if we refactor all sources to use the secret reader, so that we can remove the need for the `secrets` parameter and remove this unclear choice of which to use. Then we can change the signature to `Env(sr tpl.SecretReader)` and we no longer have this un-clarity. We can pick that up as a separate task/PR.

Reply:
> Right now, with the --ignore-missing-secrets flag the environment variables with missing secrets in them are cleared. Is this the expected behavior?

What does `are cleared` mean? Are the environment variables set with an empty value or not set at all?

The help text says:
```
      --ignore-missing-secrets   Do not return an error when a secret does not
                                 exist and use an empty value instead.
```

So I'd expect the secret reference `{{ path/to/secret }}` to be replaced with an empty string. I would expect the environment variable to be set.

Reply:
Thanks for the explanation of sources, I modified the implementation to use the `tpl.SecretReader`.

The variables are set to an empty value.